multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakEnvironmentInitalizationFilter.java

```diff
@@ -37,135 +36,146 @@
+    Claims claims = jwt.getBody();
+    LOG.debug("JWT = {}", jwt);
+    // OK, we can trust this JWT
 
     try {
-      EnvironmentContext.getCurrent().setSubject(subject);
-      Span activeSpan = tracer.activeSpan();
-      if (activeSpan != null) {
-        TracingTags.USER_ID.set(tracer.activeSpan(), subject.getUserId());
+      String username =
+          claims.get(
+              keycloakSettings.get().get(KeycloakConstants.USERNAME_CLAIM_SETTING), String.class);
+      if (username == null) { // fallback to unique id promised by spec
+        // https://openid.net/specs/openid-connect-basic-1_0.html#ClaimStability
+        username = claims.getIssuer() + ":" + claims.getSubject();
+      }
+      String id = claims.getSubject();
+
+      String email = retrieveEmail(token, claims, username, id);
+      if (email == null) {
+        throw new JwtException(
+            "Unable to authenticate user because email address is not set in keycloak profile");
       }
-      filterChain.doFilter(addUserInRequest(httpRequest, subject), response);
-    } finally {
-      EnvironmentContext.reset();
+      User user = userManager.getOrCreateUser(id, email, username);
+      return new AuthorizedSubject(
+          new SubjectImpl(user.getName(), user.getId(), token, false), permissionChecker);
+    } catch (ServerException | ConflictException e) {
+      throw new ServletException(
+          "Unable to identify user " + claims.getSubject() + " in Che database", e);
     }
   }
 
-  private HttpServletRequest addUserInRequest(
-      final HttpServletRequest httpRequest, final Subject subject) {
-    return new HttpServletRequestWrapper(httpRequest) {
-      @Override
-      public String getRemoteUser() {
-        return subject.getUserName();
-      }
+  private String retrieveEmail(String token, Claims claims, String username, String id)
+      throws ServerException {
+    String email = claims.get("email", String.class);
 
-      @Override
-      public Principal getUserPrincipal() {
-        return subject::getUserName;
+    if (isNullOrEmpty(email)) {
+      boolean userNotFound = false;
+      try {
+        userManager.getById(id);
+      } catch (NotFoundException e) {
+        userNotFound = true;
+      }
+      if (userNotFound) {
+        try {
+          EnvironmentContext.getCurrent().setSubject(new SubjectImpl(username, id, token, true));
+          Map<String, String> profileAttributes =
+              keycloakProfileRetriever.retrieveKeycloakAttributes();
+          email = profileAttributes.get("email");
+        } finally {
+          EnvironmentContext.reset();
+        }
       }
-    };
+    }
+    return email;
+  }
+
+  /** when a request came from a machine with valid token then auth is not required */
+  private boolean shouldSkipAuthentication(String token) {
+    try {
+      jwtParser.parse(token);
+      return false;
+    } catch (MachineTokenJwtException e) {
```

Comment:
can `jwtParser.parse` throw `MachineTokenJwtException` ?

Reply:
Yes, since we resolve the key using our own `KeycloakSigningKeyResolver`